sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+				AddedAt:   s.updateTime.UTC(),
+				InvoiceID: int64(s.invoice.AddIndex),
+				SetID:     setID[:],
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := s.db.UpdateAMPSubInvoiceState(
+		s.ctx, sqlc.UpdateAMPSubInvoiceStateParams{
+			SetID:        setID[:],
+			State:        int16(newState.State),
+			SettledIndex: settledIndex,
+			SettledAt:    settledAt,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Commit finalizes the update before it is written to the database. Note that
+// we don't use this directly in the SQL implementation, so the function is just
+// a stub.
+func (s *sqlInvoiceUpdater) Commit(_ invpkg.UpdateType) error {
+	return nil
+}
+
+// UpdateInvoice attempts to update an invoice corresponding to the passed
+// payment hash. If an invoice matching the passed payment hash doesn't exist
+// within the database, then the action will fail with a "not found" error.
+//
+// The update is performed inside the same database transaction that fetches the
+// invoice and is therefore atomic. The fields to update are controlled by the
+// supplied callback.
+func (i *InvoiceStore) UpdateInvoice(ctx context.Context, ref invpkg.InvoiceRef,
+	_ *invpkg.SetID, callback invpkg.InvoiceUpdateCallback) (
+	*invpkg.Invoice, error) {
+
+	var updatedInvoice *invpkg.Invoice
+
+	txOpt := InvoiceQueriesTxOptions{readOnly: false}
+	txErr := i.db.ExecTx(ctx, &txOpt, func(db InvoiceQueries) error {
+		invoice, err := i.fetchInvoice(ctx, db, ref)
+		if err != nil {
+			return err
+		}
+
+		updateTime := i.clock.Now()
+		updater := &sqlInvoiceUpdater{
+			db:         db,
+			ctx:        ctx,
+			invoice:    invoice,
+			updateTime: updateTime,
+		}
+
+		payHash := ref.PayHash()
+		updatedInvoice, err = invpkg.UpdateInvoice(
+			payHash, invoice, updateTime, callback, updater,
+		)
+
+		return err
+	})
+	if txErr != nil {
+		return updatedInvoice, txErr
+	}
+
+	return updatedInvoice, nil
+}
+
+// DeleteInvoice attempts to delete the passed invoices and all their related
+// data from the database in one transaction.
+//
+// NOTE: Settled invoices cannot be deleted with this method.
+func (i *InvoiceStore) DeleteInvoice(ctx context.Context,
+	invoicesToDelete []invpkg.InvoiceDeleteRef) error {
+
+	// All the InvoiceDeleteRef instances include the add index of the
+	// invoice. The rest was added to ensure that the invoices were deleted
+	// properly in the kv database. When we have fully migrated we can
+	// remove the rest of the fields.
+	for _, ref := range invoicesToDelete {
+		if ref.AddIndex == 0 {
+			return fmt.Errorf("unable to delete invoice using a "+
+				"ref without AddIndex set: %v", ref)
+		}
+	}
+
+	var writeTxOpt InvoiceQueriesTxOptions
+	err := i.db.ExecTx(ctx, &writeTxOpt, func(db InvoiceQueries) error {
+		for _, ref := range invoicesToDelete {
+			params := sqlc.DeleteInvoiceParams{
+				AddIndex: sqlInt64(ref.AddIndex),
+			}
+
+			if ref.SettleIndex != 0 {
```

Comment:
You mention in the comment that settled invoices should not be deleted with this method, so I think we should error out if the index is not 0 ?

Reply:
Oh that comment was actually not true. I think it was originated from the PR that this is based on. Fixed.